internal/backend/state_actions.go

```diff
@@ -46,91 +79,132 @@ func (state *State) actionUpdateMailbox(ctx context.Context, tx *ent.Tx, mboxID,
+		}
+
+		reqs = append(reqs, &DBCreateMessageReq{
+			message:    msg.Message,
+			literal:    literal,
+			body:       msg.Body,
+			structure:  msg.Structure,
+			envelope:   msg.Envelope,
+			internalID: internalID,
+		})
+	}
+
+	if _, err := DBCreateMessages(ctx, tx, reqs...); err != nil {
+		return 0, fmt.Errorf("failed to create message: %w", err)
 	}
 
-	messageUIDs, err := DBGetMessageUIDs(ctx, tx.Client(), mboxID, []string{res.ID})
+	msgIDs := []imap.InternalMessageID{internalID}
+	messageUIDs, err := DBAddMessagesToMailbox(ctx, tx, msgIDs, mboxID.InternalID)
+
 	if err != nil {
 		return 0, err
 	}
 
-	return messageUIDs[res.ID], nil
+	if err := state.forStateInMailbox(mboxID.InternalID, func(state *State) error {
+		return state.pushResponder(ctx, tx, newExists(MessageIDPair{InternalID: internalID, RemoteID: res.ID}, messageUIDs[internalID]))
+	}); err != nil {
+		return 0, err
+	}
+
+	return messageUIDs[internalID], nil
 }
 
-func (state *State) actionAddMessagesToMailbox(ctx context.Context, tx *ent.Tx, messageIDs []string, mboxID string) (map[string]int, error) {
-	var haveMessageIDs []string
+func (state *State) actionAddMessagesToMailbox(ctx context.Context,
+	tx *ent.Tx,
+	messageIDs []MessageIDPair,
+	mboxID MailboxIDPair,
+) (map[imap.InternalMessageID]int, error) {
+	var haveMessageIDs []MessageIDPair
 
-	if state.snap != nil && state.snap.mboxID == mboxID {
+	if state.snap != nil && state.snap.mboxID.InternalID == mboxID.InternalID {
 		haveMessageIDs = state.snap.getAllMessageIDs()
 	} else {
 		var err error
 
-		if haveMessageIDs, err = DBGetMailboxMessageIDs(ctx, tx.Client(), mboxID); err != nil {
+		if haveMessageIDs, err = DBGetMailboxMessageIDPairs(ctx, tx.Client(), mboxID.InternalID); err != nil {
 			return nil, err
 		}
 	}
 
-	if remMessageIDs := xslices.Filter(messageIDs, func(messageID string) bool {
+	if remMessageIDs := xslices.Filter(messageIDs, func(messageID MessageIDPair) bool {
 		return slices.Contains(haveMessageIDs, messageID)
 	}); len(remMessageIDs) > 0 {
 		if err := state.actionRemoveMessagesFromMailbox(ctx, tx, remMessageIDs, mboxID); err != nil {
 			return nil, err
 		}
 	}
 
-	if err := state.remote.AddMessagesToMailbox(state.metadataID, messageIDs, mboxID); err != nil {
+	internalIDs, remoteIDs := SplitMessageIDPairSlice(messageIDs)
+
+	if err := state.remote.AddMessagesToMailbox(ctx, state.metadataID, remoteIDs, mboxID.RemoteID); err != nil {
 		return nil, err
 	}
 
-	return state.applyMessagesAddedToMailbox(ctx, tx, mboxID, messageIDs)
+	return state.applyMessagesAddedToMailbox(ctx, tx, mboxID.InternalID, internalIDs)
 }
 
-func (state *State) actionRemoveMessagesFromMailbox(ctx context.Context, tx *ent.Tx, messageIDs []string, mboxID string) error {
-	haveMessageIDs, err := DBGetMailboxMessageIDs(ctx, tx.Client(), mboxID)
+func (state *State) actionRemoveMessagesFromMailbox(ctx context.Context, tx *ent.Tx, messageIDs []MessageIDPair, mboxID MailboxIDPair) error {
```

Comment:
does this really need to be ID pair as well? we could probably remove messages just by internalID?

also what about other actions, like `actionAddMessageFlags` -- why do we need both internal and remote ID there?

Reply:
Good point. Will Fix.

Reply:
Actually, the Remove/Add Messages needs The remote ID to process the request on the connector

Reply:
The same is also applicable for the flag variants as they end up invoking `Connector.SetMessagesSeen` and `Connector.SetMessagesFlagged`